src/Workbench.ts

```diff
@@ -145,38 +208,104 @@ export default class Workbench extends ThemeableBase<WorkbenchProperties> {
 		this.invalidate();
 	}
 
+	private _onFileTabClose = (key: string | number | undefined, label: string) => {
+		const { openFiles, onFileClose } = this.properties;
+		const idx = Number(key);
+		if (onFileClose && openFiles && openFiles[idx]) {
+			if (!this._fileTreeOpen && openFiles.length === 1) {
+				this._onToggleFiles();
+			}
+			onFileClose(openFiles[idx]);
+		}
+	}
+
+	private _onFileTabSelect = (key: string | number | undefined, label: string) => {
+		const { openFiles, onFileSelect } = this.properties;
+		const idx = Number(key);
+		if (onFileSelect && openFiles && openFiles[idx]) {
+			onFileSelect(openFiles[idx]);
+		}
+	}
+
+	private _onRun() {
+		const { onRun } = this.properties;
+		if (!this._runnerOpen) {
+			this._onToggleRunner();
+		}
+		onRun && onRun();
+	}
+
+	private _onToggleFiles() {
+		this._fileTreeOpen = !this._fileTreeOpen;
+		this._layoutEditor = true;
+		this.invalidate();
+	}
+
+	private _onToggleRunner() {
+		this._runnerOpen = !this._runnerOpen;
+		this._layoutEditor = true;
+		this.invalidate();
+	}
+
+	constructor() {
+		super();
+		// we need to "trap" window resize events so we can layout the editor intelligently
+		window.addEventListener('resize', () => {
+			this._layoutEditor = true;
+			this.invalidate();
+		});
+		window.addEventListener('beforeunload', (evt) => {
+			if (this.properties.model) {
+				evt.returnValue = 'Do you wish to navigate away from this page?';
+				return evt.returnValue;
+			}
+		});
+	}
```

Comment:
Do we need to clean up these event listeners when the workbench is unloaded?

Reply:
In the 0.2.0 there are better lifecycle methods that would make this possible, so let's open an issue to clean that up when we move to 0.2.0 (which will be my next PR).